### DISCUSSION TITLE
Use CLI train function to train on data already in memory

### DISCUSSION THREAD

**Discussion by jricheimer** [explosion/spaCy#8525] [Help: Coding & Implementations] [answered]
I'm trying to use spacy (3.0) to train a textcat model on data that is received from external systems and (ideally) will never be saved to the disk.

In other words, I have a script that will listen for messages (which contain training data) from another machine until it has enough to train a model. The training data is now already loaded in memory and in the correct format to be trained with. I don't want to have to save the data to disk and then call `spacy train` to read it. I know I can use the Internal API to create my own training loop, but I'd rather still have the convenience of the config file to set/change all the other settings.

Is there a convenient way to do what I'm trying to do?

**Comment by adrianeboyd** [comment, marked as answer]
The solution is probably to write a custom corpus reader that reads and provides the shared data to the training steps: https://spacy.io/api/top-level#corpus-readers

If you use `max_epochs = -1` (https://spacy.io/usage/training#custom-code-readers-batchers), remember that you may want to implement repeating and/or shuffling in the reader if required.

**Reply by jricheimer** [reply]
So, if I understand, I can do something like this:

```
data = wait_for_external_data()    # data is now a list of Examples

@spacy.registry.readers('read_this_data')
def reader(nlp):
    while True:
        for sample in data:
            yield sample

nlp = init_nlp(config)
spacy.training.loop.train(nlp, out_path)
```

And then in the config file, put:
```
[corpora.train]
@readers = "read_this_data"
```

Is that roughly what you are saying?

**Reply by adrianeboyd** [reply]
The config looks correct and the reader is close. You'd provide the custom reader with `spacy train -c code.py`. There's no need to do anything related to init/training in the additional code, and be aware that the registered reader function needs to return a `Callable`, not yield the batches directly.

Look at the (admittedly silly) even/odd example in this section, in particular the type signatures (the return signature for the registered function is `Callable[[Language], Iterable[Example]]`): https://spacy.io/usage/training#custom-code-readers-batchers

**Reply by jricheimer** [reply]
Maybe I'm misunderstanding. How can I provide the custom reader with a `code.py` file if the data is not accessible outside of the current python session?

**Reply by svlandeg** [reply]
If you're not calling `spacy train` from the commandline, you should be able to just import the custom reader into your code, and it'll be available in spaCy's registry.

**Reply by jricheimer** [reply]
Thanks for the tip @svlandeg.
@adrianeboyd The approach I summarized above did work (after I made the registered function return a callable as you noted). Thank you for your help.

**Reply by adrianeboyd** [reply]
Ah, I originally thought there would be multiple processes. Glad to hear it's working!
